### PR TITLE
Introduce ProcessGroupCudaP2P

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126599
* #126598
* #126634
* __->__ #122163

## Context
This stack prototypes automatic micro-pipelining of `all-gather -> matmul` and `matmul -> reduce-scatter` via Inductor. The idea originates from the paper [Overlap Communication with Dependent Computation via
Decomposition in Large Deep Learning Models](https://dl.acm.org/doi/pdf/10.1145/3567955.3567959). The implementation and some key optimizations are heavily influenced by @lw's implementation in xformers.

The stack contains several components:
- `ProcessGroupCudaP2P` - a thin wrapper around `ProcessGroupNCCL`. It in addition maintains a P2P workspace that enables SM-free, one-sided P2P communication which is needed for optimal micro-pipelining.
- `fused_all_gather_matmul` and `fused_matmul_reduce_scatter` dispatcher ops.
- Post-grad fx pass that detects `all-gather -> matmul` and `matmul -> reduce-scatter` and replaces them with the fused dispatcher ops.

To enable the prototype feature:
- Set the distributed backend to `cuda_p2p`.
- Set `torch._inductor.config._micro_pipeline_tp` to `True`.

*NOTE: the prototype sets nothing in stone w.r.t to each component's design. The purpose is to have a performant baseline with reasonable design on which each component can be further improved.*

## Benchmark
Setup:
- 8 x H100 (500W) + 3rd gen NVSwitch.
- Llama3 8B training w/ torchtitan.
- 8-way TP. Reduced the number of layers from 32 to 8 for benchmarking purpose.

Trace (baseline): https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmpjaz8zgx0
image

Trace (w/ micro pipelining): https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmpn073b4wn
image

## This PR
`ProcessGroupCudaP2P` is a thin wrapper around `ProcessGroupNCCL`. By default, it routes all collectives to the underlying `ProcessGroupNCCL`. In addition, `ProcessGroupCudaP2P` initializes a P2P workspace that allows direct GPU memory access among the members. The workspace can be used in Python to optimize intra-node communication patterns or to create custom intra-node collectives in CUDA.

`ProcessGroupCudaP2P` aims to bridge the gap where certain important patterns can be better optimized via fine-grained P2P memory access than with collectives in the latest version of NCCL. It is meant to complement NCCL rather than replacing it.
Usage:
```
    # Using ProcessGroupCudaP2P
    dist.init_process_group(backend="cuda_p2p", ...)

    # Using ProcessGroupCudaP2P while specifying ProcessGroupCudaP2P.Options
    pg_options = ProcessGroupCudaP2P.Options()
    dist.init_process_group(backend="cuda_p2p", pg_options=pg_options, ...)

    # Using ProcessGroupCudaP2P while specifying ProcessGroupNCCL.Options
    pg_options = ProcessGroupNCCL.Options()
    dist.init_process_group(backend="cuda_p2p", pg_options=pg_options, ...)

    # Using ProcessGroupCudaP2P while specifying both
    # ProcessGroupCudaP2P.Options and ProcessGroupNCCL.Options
    pg_options = ProcessGroupCudaP2P.Options()
    pg_options.nccl_options = ProcessGroupNCCL.Options()
    dist.init_process_group(backend="cuda_p2p", pg_options=pg_options, ...)

    # Down-casting the backend to access p2p buffers for cuda_p2p specific
    # optimizations
    if is_cuda_p2p_group(group):
        backend = get_cuda_p2p_backend(group)
        if required_p2p_buffer_size > backend.get_buffer_size():
            # fallback
        p2p_buffer = backend.get_p2p_buffer(...)
    else:
        # fallback
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

### CODE REVIEW

File: torch/distributed/_cuda_p2p/__init__.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+def get_cuda_p2p_backend(group: c10d.ProcessGroup) -> "c10d.ProcessGroupCudaP2P":
+    if not is_cuda_p2p_group(group):
+        raise TypeError("group is not a cuda_p2p process group.")
+    return cast(
+        c10d.ProcessGroupCudaP2P,
+        group._get_backend(torch.device("cuda")),
+    )
+
+
+def get_p2p_buffer_size(group: c10d.ProcessGroup) -> int:
```
Comment (wanchaol) [inline torch/distributed/_cuda_p2p/__init__.py:111]: Wondering why we need this method as I see no callsites to it?

File: torch/csrc/distributed/c10d/ProcessGroupCudaP2P.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+      options->nccl_options);
+  nccl_backend_->setSequenceNumberForGroup();
+
+  p2p_backend_ = c10::make_intrusive<IntraNodeComm>(
+      c10::make_intrusive<PrefixStore>("p2p", store),
+      rank,
+      size,
+      options->buffer_size);
+  if (!p2p_backend_->rendezvous()) {
+    p2p_backend_ = nullptr;
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/ProcessGroupCudaP2P.cpp:29]: shall we throw an error here or you want to preserve the grace fallback to NCCL?

**Review by wanchaol — APPROVED** [review]
Very clean backend implementation!

File: torch/distributed/_cuda_p2p/__init__.py (hunk 1 above)
Comment (yifuwang) [reply torch/distributed/_cuda_p2p/__init__.py:111]: Hey @wanchaol, it's used in the subsequent PR. But I should add a test for it in this PR.

File: torch/csrc/distributed/c10d/ProcessGroupCudaP2P.cpp (hunk 2 above)
Comment (yifuwang) [reply torch/csrc/distributed/c10d/ProcessGroupCudaP2P.cpp:29]: Yes the intention is to gracefully fallback. `p2p_backend_` being `nullptr` will make `is_p2p_available` return false, which instructs the application layer to fallback. But I also see benefits of just erroring out here. Let me think more.

**Comment by kwen2501** [comment]
Hi @yifuwang, the trace figures definitely look exciting!

A couple of questions for me to better understand the strategy:
- what APIs does `ProcessGroupCudaP2P` override on top of `ProcessGroupNCCL`?
- what APIs does `ProcessGroupCudaP2P` want to additionally expose to uses?
- For users to use this feature, would they still call `dist.all_gather_into_tensor` and `dist.reduce_scatter_tensor`? Or they would call something else?

**Comment by jithunnair-amd** [comment]
@pytorchbot revert -m "This is breaking ROCm distributed CI on trunk" -c nosignal

Eg. 
ProcessGroupCUDAP2P

cc @huydhn Please help issue the revert if above command fails. I've added ciflow/periodic label to this PR to trigger ROCm distributed CI tests.

**Comment by yifuwang** [comment]
Fixed the rocm test failure.

@pytorchbot merge
